src/backend/distributed/executor/subplan_execution.c

```diff
@@ -81,3 +117,148 @@ ExecuteSubPlans(DistributedPlan *distributedPlan)
+		{
+			elog(DEBUG4, "subplan %s is used in all workers", resultId);
+
+			break;
+		}
+		else
+		{
+			entry->nodeList = AppendAllAccessedWorkerNodes(entry->nodeList,
+														   distributedPlan);
+
+			elog(DEBUG4, "subplan %s is used in %lu", resultId, distributedPlan->planId);
+		}
+	}
+
+	/* descend into the subPlans */
+	foreach(subPlanCell, subPlanList)
+	{
+		DistributedSubPlan *subPlan = (DistributedSubPlan *) lfirst(subPlanCell);
+		CustomScan *customScan = FetchCitusCustomScanIfExists(subPlan->plan->planTree);
+		if (customScan)
+		{
+			DistributedPlan *distributedPlanOfSubPlan = GetDistributedPlan(customScan);
+			RecordSubplanExecutionsOnNodes(intermediateResultHash,
+										   distributedPlanOfSubPlan);
+		}
+	}
+}
+
+
+/*
+ * AppendAllAccessedWorkerNodes iterates over all the tasks in a distributed plan
+ * to create the list of worker nodes that can be accessed when this plan is executed.
+ *
+ * If there are multiple placements of a Shard, all of them are considered.
+ */
+static List *
+AppendAllAccessedWorkerNodes(List *workerNodeList, DistributedPlan *distributedPlan)
+{
+	List *taskList = distributedPlan->workerJob->taskList;
+	ListCell *taskCell = NULL;
+
+	foreach(taskCell, taskList)
+	{
+		Task *task = lfirst(taskCell);
+		ListCell *placementCell = NULL;
+		foreach(placementCell, task->taskPlacementList)
+		{
+			ShardPlacement *placement = lfirst(placementCell);
+			workerNodeList = list_append_unique_int(workerNodeList, placement->nodeId);
+
+			/* early return if all the workers are accessed */
+			if (list_length(workerNodeList) == GetWorkerNodeCount())
+			{
+				return workerNodeList;
+			}
+		}
+	}
+
+	return workerNodeList;
+}
+
+
+/*
+ * MakeIntermediateResultHTAB is a helper method that creates a Hash Table that store information on the intermediate result.
+ */
+static HTAB *
+MakeIntermediateResultHTAB()
+{
+	HTAB *intermediateResultHash = NULL;
+	uint32 hashFlags = 0;
+	HASHCTL info = { 0 };
+	int initialNumberOfElements = 64;
+
+	info.keysize = sizeof(IntermediateResultHashKey);
+	info.entrysize = sizeof(IntermediateResultHashEntry);
+	info.hash = string_hash;
+	info.hcxt = CurrentMemoryContext;
+	hashFlags = (HASH_ELEM | HASH_FUNCTION | HASH_CONTEXT);
+
+	intermediateResultHash = hash_create("Intermediate results hash",
+										 initialNumberOfElements, &info, hashFlags);
+
+	return intermediateResultHash;
+}
+
+
+/**
+ * SearchIntermediateResult searches for an entry in a hash table and creates a
+ * new entry if the search key is not present.
+ *
+ */
+static IntermediateResultHashEntry *
+SearchIntermediateResult(HTAB *intermediateResultHash, char *resultId, bool *found)
+{
+	IntermediateResultHashKey key;
+	IntermediateResultHashEntry *entry = NULL;
+
+	strlcpy(key.intermediate_result_id, resultId, NAMEDATALEN);
+	entry = hash_search(intermediateResultHash, &key, HASH_ENTER, found);
+
+	/* if the key was not found, update the entry with sane defaults */
+	if (!found)
```

Comment:
if (!found) > if (!*found)

Reply:
The variable is declared as `bool found = false;` above. So I think this one is actually correct.